### PR TITLE
test: Build fuzz targets into seperate executables

### PR DESCRIPTION
**Pull request by maflcko**
Currently our fuzzer is a single binary that decides on the first few bits of the buffer what target to pick. This is ineffective as the fuzzer needs to "learn" how the fuzz targets are organized and could get easily confused. Not to mention that the (seed) corpus can not be categorized by target, since targets might "leak" into each other. Also the corpus would potentially become invalid if we ever wanted to remove a target...

Solve that by building each fuzz target into their own executable.

### CODE REVIEW

**Comment by practicalswift** [comment]
Concept ACK

Thanks for doing this!

File: src/Makefile.test.include
Comment (practicalswift) [inline src/Makefile.test.include]: Use `.h` instead of `.hpp` :-)

File: src/test/test_bitcoin_fuzzy-blockheader_deserialize.cpp
Comment (practicalswift) [inline src/test/test_bitcoin_fuzzy-blockheader_deserialize.cpp]: Drop the unused `e` variable. Applies throughout this PR.

File: src/test/test_bitcoin_fuzzy-transaction_deserialize.cpp
Comment (practicalswift) [inline src/test/test_bitcoin_fuzzy-transaction_deserialize.cpp]: Make `buffer` const reference. Applies throughout this PR :-)

File: src/test/test_bitcoin_fuzzy.hpp
Comment (practicalswift) [inline src/test/test_bitcoin_fuzzy.hpp]: Nit: Could drop unused variable names `argc` and `argv`.

Comment (fanquake) [reply src/test/test_bitcoin_fuzzy.hpp]: Related: https://llvm.org/docs/LibFuzzer.html#startup-initialization

**Comment by laanwj** [comment]
Concept ACK

I think this should be behind a `configure` flag, building so many executables is going to be slow when statically linking, or with slow filesystems, and it contributes nothing to testing unless the user is planning to do fuzzing.

**Comment by cjdelisle** [comment]
As a non-stakeholder, feel free to ignore, but as someone who is using the same test methodology, I would like to understand why you're proposing to change.

I have a number of reasons why I tend to prefer a single fuzz entry point, but I would like to know if there is any evidence that putting the test-case in the data can harm the fuzzer's ability to efficiently find paths.

My reasoning is as follows:

1. Personally, I find it better to allow someone with a large machine to be able to run a simple command and begin fuzzing with minimal effort, if each case is fuzzed separately then this is not really practical.
2. Secondly, I personally am a big fan of maximum fuzz coverage, which means any test that can plausibly make use of randomized data should be accessible from the fuzzer. So clearly I would not want to require people to launch each test separately.
3. It seems like if one would like to fuzz a particular test case, their need can be easily supported by allowing them to pass an argument to the running which causes it to `return 0` if the test is specified to anything else, thus the fuzzer will quickly prune attempts which touch other test cases.
4. If there is a large corpus of generated test data, it seems that one could use a simple python script to rename the test files to contain the name of the test case which they're touching, solving the concern about organization.

Thanks,
Caleb

**Comment by Crypt-iQ** [comment]
I think that the fuzzing targets should be run on the corpus.  Maybe this can be done in another PR.
https://github.com/bitcoin/bitcoin/issues/11045#issuecomment-334516409
https://github.com/google/oss-fuzz/blob/master/docs/ideal_integration.md#regression-testing

File: doc/fuzzing.md
Comment (Crypt-iQ) [inline doc/fuzzing.md]: User is already in the `src` directory so this will fail

**Comment by laanwj** [comment]
> I think that the fuzzing targets should be run on the corpus as part of regression testing. This would require the corpus to be included in this project. Is there a reason why it's not currently included? 

The idea is to have a separate repository with the corpus. The problem with including it in the main repository, besides taking up space, is that e.g. all changes have to go through the bottleneck of maintainers here.

**Comment by laanwj** [comment]
Probably want to remove the links to test inputs from `doc/fuzzing.md` here, as they'll no longer work as-is and one of the links is broken (#15028).

File: doc/fuzzing.md
Comment (MarcoFalke) [reply doc/fuzzing.md]: Done (changed it to `make`)

**Comment by MarcoFalke** [comment]
@laanwj The previous corpus can still be used by stripping the first few (4?) bytes off of the seeds. I will do that (and move them to a separate repo) after this pull has been merged.

**Comment by MarcoFalke** [comment]
This hasn't received any review, but I am going to merge it tomorrow unless someone objects to that.

**Comment by sipa** [comment]
Wouldn't it be easier to accomplish this by having a since source file for all fuzz target, but with a macro define (-D... argument) to select which fuzzer is invoked by main?

**Comment by MarcoFalke** [comment]
@sipa Wouldn't that require to pass in and compile for each fuzz target? That seems to move the verbosity to the user/ci script/fuzz runner.

Also having a single source file makes it harder to see what each single fuzz test does (and depends on).

**Comment by sipa** [comment]
@MarcoFalke The build system would automatically build all of them, supplying the necessary -D arguments for each binary.

**Comment by MarcoFalke** [comment]
Now split into two commits, where the top commit is half move-only:

```
git diff 02606835e6616af~ 02606835e6616af --color-moved=dimmed-zebra src/test

File: configure.ac
Comment (practicalswift) [inline configure.ac]: Bikeshedding perhaps but IMHO `--enable-fuzzing` feels more natural than `--enable-fuzz`.

File: src/Makefile.test.include
Comment (ryanofsky) [inline src/Makefile.test.include]: You could deduplicate some boilerplate by defining common variables like:

```
fuzz_common_ldflags = $(RELDFLAGS) $(AM_LDFLAGS) $(LIBTOOL_APP_LDFLAGS)
fuzz_common_ldadd = $(LIBUNIVALUE) $(LIBBITCOIN_SERVER) $(LIBBITCOIN_COMMON)
```

and then individual targets could be shortened to:

```
test_fuzz_transaction_deserialize_LDFLAGS = $(fuzz_common_ldflags)
test_fuzz_transaction_deserialize_LDADD = $(fuzz_common_ldadd)
...

test_fuzz_block_deserialize_LDFLAGS = $(fuzz_common_ldflags)
test_fuzz_block_deserialize_LDADD = $(fuzz_common_ldadd)
...
```

At least this is what I did in 2060a30063866ed1599134fe28846a69deda773c for #10102. One catch is that LDFLAGS, LDADD, etc suffix can't be capitalized or the variables will be treated specially by automake.

File: src/test/test_bitcoin_fuzzy.cpp
Comment (ryanofsky) [inline src/test/test_bitcoin_fuzzy.cpp]: I think it would make things less complicated just to use separate source files, rather than using these preprocessor defines. These defines don't seem to really decrease duplication, but I guess they they do have the advantage of making it easy to see the different test cases all in one file.

**Review by ryanofsky — APPROVED** [review]
utACK 2ca632e5b44a8385989c8539cc4e30e60fdee16c

File: src/test/test_bitcoin_fuzzy.cpp
Comment (MarcoFalke) [reply src/test/test_bitcoin_fuzzy.cpp]: This was my initial solution, but I changed it back to minimize the diff

**Pull request merged**
